compiler/quilt/tools/command.py

```diff
@@ -879,84 +887,125 @@ def install(package, hash=None, version=None, tag=None, force=False):
+                                    if not response.ok:
+                                        message = "Download failed for %s:\nURL: %s\nStatus code: %s\nResponse: %r\n" % (
+                                            obj_hash, response.request.url, response.status_code, response.text
+                                        )
+                                        with lock:
+                                            tqdm.write(message)
+                                        break
+
+                                    # Fragments have the 'Content-Encoding: gzip' header set to make requests ungzip
+                                    # them automatically - but that turned out to be a bad idea because it makes
+                                    # resuming downloads impossible.
+                                    # HACK: For now, just delete the header. Eventually, update the data in S3.
+                                    response.raw.headers.pop('Content-Encoding', None)
+
+                                    # Make sure we're getting the expected range.
+                                    content_range = response.headers.get('Content-Range', '')
+                                    match = CONTENT_RANGE_RE.match(content_range)
+                                    if not match or not int(match.group(1)) == starting_length:
+                                        with lock:
+                                            tqdm.write("Unexpected Content-Range: %s" % content_range)
+                                        break
+
+                                    for chunk in response.iter_content(CHUNK_SIZE):
+                                        output_file.write(chunk)
+
+                                success = True
+                                break  # Done!
+                            except requests.exceptions.ConnectionError as ex:
+                                if attempt < S3_TIMEOUT_RETRIES - 1:
+                                    with lock:
+                                        tqdm.write("Download for %s timed out; retrying..." % obj_hash)
+                                else:
+                                    with lock:
+                                        tqdm.write("Download failed for %s: %s" % (obj_hash, ex))
+                                break
 
-            # Check the hash of the result.
-            file_hash = digest_file(temp_path)
-            if file_hash != download_hash:
-                os.remove(temp_path)
-                raise CommandException("Fragment hashes do not match: expected %s, got %s." %
-                                       (download_hash, file_hash))
+                    if not success:
+                        # We've already printed an error, so not much to do - just move on to the next object.
+                        continue
+
+                    # Ungzip the downloaded fragment.
+                    temp_path = store.temporary_object_path(obj_hash)
```

Comment:
Maybe some of this code can move into PackageStore?

Reply:
Which part of it? The ungzipping code?